File: backends/node/src/env.ts
export interface Env {
  PORT: number;
  DATABASE_URL: string;
  SMTP_HOST: string;
  SMTP_PORT: number;
  SMTP_FROM: string;
  HEARTH_PUBLIC_BASE_URL: string;
}

function required(name: string): string {
  const value = process.env[name];
  if (!value) {
    throw new Error(`Missing required env var: ${name}`);
  }
  return value;
}

export function loadEnv(): Env {
  return {
    PORT: parseInt(process.env.PORT ?? '5001', 10),
    DATABASE_URL: required('DATABASE_URL'),
    SMTP_HOST: process.env.SMTP_HOST ?? 'localhost',
    SMTP_PORT: parseInt(process.env.SMTP_PORT ?? '1025', 10),
    SMTP_FROM: process.env.SMTP_FROM ?? 'hearth@localhost',
    HEARTH_PUBLIC_BASE_URL: process.env.HEARTH_PUBLIC_BASE_URL ?? 'http://localhost:3000',
  };
}

File: backends/python/src/hearth_python/env.py
from __future__ import annotations

import os
from dataclasses import dataclass


@dataclass(frozen=True)
class Env:
    port: int
    database_url: str
    smtp_host: str
    smtp_port: int
    smtp_from: str
    hearth_public_base_url: str


def _required(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        raise RuntimeError(f"Missing required env var: {name}")
    return value


def load_env() -> Env:
    return Env(
        port=int(os.environ.get("PORT", "5003")),
        database_url=_required("DATABASE_URL"),
        smtp_host=os.environ.get("SMTP_HOST", "localhost"),
        smtp_port=int(os.environ.get("SMTP_PORT", "1025")),
        smtp_from=os.environ.get("SMTP_FROM", "hearth@localhost"),
        hearth_public_base_url=os.environ.get(
            "HEARTH_PUBLIC_BASE_URL", "http://localhost:3000"
        ),
    )
